File: src/console.ts
import pc from "picocolors";
import type { LogLevel, LogRecord } from "./types.ts";

const LEVEL_COLORS: Record<LogLevel, (s: string) => string> = {
	trace: pc.gray,
	debug: pc.cyan,
	info: pc.green,
	warn: pc.yellow,
	error: pc.red,
	fatal: pc.bgRed,
};

// Save original console methods so printLogRecord always bypasses any overrides
export const originalConsole: {
	[K in "log" | "warn" | "error" | "debug" | "info" | "trace"]: (...args: unknown[]) => void;
} = {
	log: console.log.bind(console),
	warn: console.warn.bind(console),
	error: console.error.bind(console),
	debug: console.debug.bind(console),
	info: console.info.bind(console),
	trace: console.trace.bind(console),
};

export function formatLogRecord(record: LogRecord): string {
	const time = record.timestamp.slice(11, 23);
	const colorFn = LEVEL_COLORS[record.level];
	const levelStr = colorFn(record.level.toUpperCase().padEnd(5));
	const svc = record.service ? `[${pc.blue(record.service)}] ` : "";
	const projParts = [record.project, record.branch].filter(Boolean).join("@");
	const ver = record.version ? `v${record.version}` : "";
	const projLabel = [projParts, ver].filter(Boolean).join(" ");
	const proj = projLabel ? `[${pc.magenta(projLabel)}] ` : "";
	const duration =
		record.meta && typeof record.meta.duration_ms === "number"
			? ` ${pc.dim(`(${record.meta.duration_ms}ms)`)}`
			: "";
	const meta =
		record.meta && Object.keys(record.meta).length > 0
			? ` ${pc.dim(safeStringify(record.meta))}`
			: "";
	return `${pc.dim(time)} ${levelStr} ${proj}${svc}${record.message}${duration}${meta}`;
}

/**
 * `JSON.stringify` throws on BigInt and on circular references — both of
 * which leak into log meta when callers pass DB rows or driver objects
 * directly. A throw here aborts the local console formatter and the user
 * loses the log entirely. Coerce BigInt to string in a replacer; on any
 * remaining error (cycles, getters that throw), fall back to a marker so
 * the rest of the line still renders.
 */
function safeStringify(value: unknown): string {
	try {
		return JSON.stringify(value, (_k, v) => (typeof v === "bigint" ? v.toString() : v));
	} catch {
		return "[unserializable meta]";
	}
}

export function printLogRecord(record: LogRecord): void {
	const formatted = formatLogRecord(record);
	if (record.level === "error" || record.level === "fatal") {
		originalConsole.error(formatted);
	} else if (record.level === "warn") {
		originalConsole.warn(formatted);
	} else {
		originalConsole.log(formatted);
	}
}

File: python/src/relog/_console.py
from __future__ import annotations

import json
import sys
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ._types import LogLevel, LogRecord

_RESET = "\033[0m"
_DIM = "\033[2m"
_RED = "\033[31m"
_GREEN = "\033[32m"
_YELLOW = "\033[33m"
_BLUE = "\033[34m"
_MAGENTA = "\033[35m"
_CYAN = "\033[36m"
_GRAY = "\033[90m"
_BG_RED = "\033[41m"

_LEVEL_COLORS: dict[str, str] = {
    "trace": _GRAY,
    "debug": _CYAN,
    "info": _GREEN,
    "warn": _YELLOW,
    "error": _RED,
    "fatal": _BG_RED,
}

_use_colors: bool | None = None


def _colors_enabled() -> bool:
    global _use_colors
    if _use_colors is None:
        _use_colors = hasattr(sys.stdout, "isatty") and sys.stdout.isatty()
    return _use_colors


def _color(code: str, text: str) -> str:
    if not _colors_enabled():
        return text
    return f"{code}{text}{_RESET}"


def _dim(text: str) -> str:
    return _color(_DIM, text)


def format_log_record(record: LogRecord) -> str:
    time = record.timestamp[11:23]
    level_color = _LEVEL_COLORS.get(record.level, "")
    level_str = _color(level_color, record.level.upper().ljust(5))

    svc = f"[{_color(_BLUE, record.service)}] " if record.service else ""

    proj_parts = "@".join(filter(None, [record.project, record.branch]))
    ver = f"v{record.version}" if record.version else ""
    proj_label = " ".join(filter(None, [proj_parts, ver]))
    proj = f"[{_color(_MAGENTA, proj_label)}] " if proj_label else ""

    duration = ""
    if record.meta and isinstance(record.meta.get("duration_ms"), (int, float)):
        duration = f" {_dim(f'({record.meta["duration_ms"]}ms)')}"

    meta = ""
    if record.meta and len(record.meta) > 0:
        meta = f" {_dim(json.dumps(record.meta, default=str))}"

    return f"{_dim(time)} {level_str} {proj}{svc}{record.message}{duration}{meta}"


def print_log_record(record: LogRecord) -> None:
    formatted = format_log_record(record)
    if record.level in ("error", "fatal"):
        print(formatted, file=sys.stderr)
    elif record.level == "warn":
        print(formatted, file=sys.stderr)
    else:
        print(formatted)
